### PR TITLE
coverage: Use a query to identify which counter/expression IDs are used

### PR DESCRIPTION
**Pull request by Zalathar**
Given that we already have a query to identify the highest-numbered counter ID in a MIR body, we can extend that query to also build bitsets of used counter/expression IDs. That lets us avoid some messy coverage bookkeeping during the main MIR traversal for codegen.

This does mean that we fail to treat some IDs as used in certain MIR-inlining scenarios, but I think that's fine, because it means that the results will be consistent across all instantiations of a function.

---

There's some more cleanup I want to do in the function coverage collector, since it isn't really collecting anything any more, but I'll leave that for future work.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs [hunk 1]
```diff
@@ -192,23 +195,23 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
...
                     .try_get_mcdc_condition_bitmap(&instance, decision_depth)
                     .expect("mcdc cond bitmap should have been allocated for updating");
                 let cond_index = bx.const_i32(index as i32);
                 bx.mcdc_condbitmap_update(cond_index, cond_bitmap);
             }
             CoverageKind::TestVectorBitmapUpdate { bitmap_idx, decision_depth } => {
-                drop(coverage_map);
-                let cond_bitmap = bx.coverage_cx()
-                                    .try_get_mcdc_condition_bitmap(&instance, decision_depth)
-                                    .expect("mcdc cond bitmap should have been allocated for merging into the global bitmap");
```
Comment (Zalathar) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs:211]: For some reason this long string causes rustfmt to completely give up on the enclosing statement, so I've split it up to work around that.

**Comment by Zalathar** [comment]
Rebased due to import conflicts; no changes.

**Comment by fee1-dead** [comment]
not familiar with coverage, sorry

r? compiler

**Comment by Nadrieril** [comment]
not familiar with coverage either, sorry

r? compiler

**Comment by Zalathar** [comment]
Rebased again; no changes.

**Pull request merged**